src/SAWScript/Crucible/MIR/Builtins.hs

```diff
@@ -26,5 +210,952 @@ mir_load_module inputFile = do
+      , (natValue (knownNat @16),  B16View)
+      , (natValue (knownNat @32),  B32View)
+      , (natValue (knownNat @64),  B64View)
+      , (natValue (knownNat @128), B128View)
+      ]
+
+-- | Convert a 'Mir.Substs' value to a 'SubstsView' value.
+substsView :: Mir.Substs -> SubstsView
+substsView (Mir.Substs tys) = SubstsView (map tyView tys)
+
+-- | Convert a 'Mir.FnSig' value to a 'FnSigView' value.
+fnSigView :: Mir.FnSig -> FnSigView
+fnSigView (Mir.FnSig argTys retTy abi spreadarg) =
+  FnSigView (map tyView argTys) (tyView retTy) abi spreadarg
+
+-- | Returns the Cryptol type of a MIR type, returning 'Nothing' if it is not
+-- easily expressible in Cryptol's type system or if it is not currently
+-- supported.
+cryptolTypeOfActual :: Mir.Ty -> Maybe Cryptol.Type
+cryptolTypeOfActual mty =
+  case mty of
+    Mir.TyBool      -> Just Cryptol.tBit
+    Mir.TyChar      -> Just $ Cryptol.tWord $ Cryptol.tNum (32 :: Integer)
+    Mir.TyUint bs   -> baseSizeType bs
+    Mir.TyInt  bs   -> baseSizeType bs
+    Mir.TyArray t n -> Cryptol.tSeq (Cryptol.tNum n) <$> cryptolTypeOfActual t
+    Mir.TyTuple tys -> Cryptol.tTuple <$> traverse cryptolTypeOfActual tys
+
+    Mir.TyFloat _      -> Nothing
+    Mir.TyStr          -> Nothing
+    Mir.TyAdt _ _ _    -> Nothing
+    Mir.TyRef _ _      -> Nothing
+    Mir.TyFnDef _      -> Nothing
+    Mir.TyFnPtr _      -> Nothing
+    Mir.TyRawPtr _ _   -> Nothing
+    Mir.TyClosure _    -> Nothing
+    Mir.TySlice _      -> Nothing
+    Mir.TyDowncast _ _ -> Nothing
+    Mir.TyNever        -> Nothing
+    Mir.TyForeign      -> Nothing
+    Mir.TyLifetime     -> Nothing
+    Mir.TyConst        -> Nothing
+    Mir.TyErased       -> Nothing
+    Mir.TyInterned _   -> Nothing
+    Mir.TyDynamic _    -> Nothing
+  where
+    baseSizeType :: Mir.BaseSize -> Maybe Cryptol.Type
+    baseSizeType Mir.B8    = Just $ Cryptol.tWord $ Cryptol.tNum (8 :: Integer)
+    baseSizeType Mir.B16   = Just $ Cryptol.tWord $ Cryptol.tNum (16 :: Integer)
+    baseSizeType Mir.B32   = Just $ Cryptol.tWord $ Cryptol.tNum (32 :: Integer)
+    baseSizeType Mir.B64   = Just $ Cryptol.tWord $ Cryptol.tNum (64 :: Integer)
+    baseSizeType Mir.B128  = Just $ Cryptol.tWord $ Cryptol.tNum (128 :: Integer)
+    baseSizeType Mir.USize = Just $ Cryptol.tWord $ Cryptol.tNum $ natValue $ knownNat @Mir.SizeBits
+
+-- | Given a function name @fnName@, attempt to look up its corresponding
+-- 'Mir.DefId'. Currently, the following types of function names are permittd:
+--
+-- * @<crate_name>/<disambiguator>::<function_name>: A fully disambiguated name.
+--
+-- * @<crate_name>::<function_name>: A name without a disambiguator. In this
+--   case, SAW will attempt to look up a disambiguator from the @crateDisambigs@
+--   map. If none can be found, or if there are multiple disambiguators for the
+--   given @<crate_name>@, then this will fail.
+findDefId :: Map Text (NonEmpty Text) -> Text -> TopLevel Mir.DefId
+findDefId crateDisambigs fnName = do
```

Comment:
This function implements the MIR identifier lookup logic described in the SAW manual.